Automattic/Gravatar-SDK-Android
# QuickEditor: Error handling

Closes #262

### Description
This PR adds better error handling in the QuickEditor. 

1. I've modified the `QuickEditorError` types to reflect new changes https://github.com/Automattic/Gravatar-SDK-Android/commit/425c8548f02b3d71417c0f2c6bd4a19299b152d9.
2. `AvatarRepository.getAvatars` was modified to use `catching` versions so that we don't have to copy/expose the Exception mapping to `ErrorType` from the core module https://github.com/Automattic/Gravatar-SDK-Android/commit/425c8548f02b3d71417c0f2c6bd4a19299b152d9.
3. `QEButton` https://github.com/Automattic/Gravatar-SDK-Android/commit/8c4ef64b6d761933a1535e268dfa1a602b117cfb, `QESectionTitle` https://github.com/Automattic/Gravatar-SDK-Android/commit/d79600308f591c2191b4289222bdca3d101795d3, and `QESectionMessage` https://github.com/Automattic/Gravatar-SDK-Android/commit/5070bf838304cb03939d9a80ba723304fca462fe were created to reuse between different sections.
4. `ErrorSection` was added and implemented in the `AvatarPicker` https://github.com/Automattic/Gravatar-SDK-Android/commit/d8af7985135c9022f534e205e919562f04898255.
5. Added a dynamic height to the loading composable to avoid bottom sheet height adjustments when retrying network requests https://github.com/Automattic/Gravatar-SDK-Android/commit/74b5258b0d1503a2356afadca86052b91cf579d2.

Here are the currently implemented error states:

| No internet | Server error (500+) | Request error |
|---|---|---|
| <img width="692" alt="Screenshot 2024-08-21 at 14 42 33" src="https://github.com/user-attachments/assets/43daccf9-7c64-4784-aefe-00b593c6cb7f"> | <img width="692" alt="Screenshot 2024-08-21 at 14 44 32" src="https://github.com/user-attachments/assets/277204a9-6424-4ded-b4af-e4be91eda896"> | <img width="692" alt="Screenshot 2024-08-21 at 14 45 29" src="https://github.com/user-attachments/assets/28c3a371-ff8c-4338-ae91-32baf5fdeb7f"> |

### Testing Steps

**Important Note**: The backend is currently returning the wrong format for the `updated date` field. As a workaround, you can apply this patch to comment that field from the Avatar object.
[Gravatar-Android-13-04-48.patch](https://github.com/user-attachments/files/16455284/Gravatar-Android-13-04-48.patch)

2. Run the DemoApp 
3. Go to Avatar Update tab
4. Tap `Update Avatar` button
5. Follow the steps to log in (if necessary)
6. Close the QE
7. Enable airplane mode
8. Open the QE  
9. Confirm you see No internet error
10. Close the QE
11. Enable request mocking (HTTP toolkit, charles or what ever tool you use) 
13. Fake `/avatars` endpoint to return 500
14. Open the QR
15. Confirm you see error about server problems
16. Close the QE
13. Fake `/avatars` endpoint to return 404
14. Open the QR
15. Confirm you see general error message
16. Close the QE
17. Smoke test the QE by selecting/uploading an avatar.
